### ISSUE TITLE
Change cursor datatype in RaceBuffer to u64 (and dealing with the race conditions this creates)

### ISSUE THREAD

**Issue by mullr** [auxoncorp/modality-probe#187]
We've identified a troublesome race condition with the current overflow handling code; a change to `u64` eliminates this for all practical implementations, and allow us to eliminate overflow handling.

**Comment by nathanaschbacher** [comment]
Updating them to be U64s makes them multi-word writes on most hardware we expect to deal with, which makes them potentially interruptable mid-write if I understand things correctly. Doesn't that mean that something like the debug reader might end up reading a value from the cursor location that's half-written?

Is there the possibility of a solution to managing the cursors in a way that doesn't depend on numbers that need to grow forever by-design, but can't due to wrap around in-practice? Or maybe something that maintains other single-word state that's tracking the relationship or epochs of the relationships between the cursors?

**Comment by mullr** [comment]
> Updating them to be U64s makes them multi-word writes on most hardware we expect to deal with, which makes them potentially interruptable mid-write if I understand things correctly. Doesn't that mean that something like the debug reader might end up reading a value from the cursor location that's half-written?

Yes. We've discussed several ways to accommodate this based on the fact that these number aren't arbitrarily set, they're incremented. For example: read the high word, then the low word, then the high word again. If the high word changed, then you know the low word must have rolled over to zero in the interim. There are a few different variations of that. 

But: the intent of this ticket is to allow us to get the RaceBuffer merged into mainline modality-probe for regular (non-debug-reader) use, unblocking a bunch of other work which depends on having a working ring buffer in place. It's possible that none of the debug-reader strategies will actually work out, and we have to do something different; in that case we'll figure something out. But we will still have something in place for local-operation, which is the short-term priority. 

> Is there the possibility of a solution to managing the cursors in a way that doesn't depend on numbers that need to grow forever by-design, but can't due to wrap around in-practice? Or maybe something that maintains other single-word state that's tracking the relationship or epochs of the relationships between the cursors?

Yes, of course there is the possibility. But the design constraints of read-only access (all reader state is local to the reader) and being able to tell how many events were overwritten between reads leaves me at a loss.
